### PR TITLE
`explicit_deref_methods` improvements 

### PR DESCRIPTION
**Pull request by Jarcho**
Breaking up #6837 

changelog: `explicit_deref_methods` will lint chained `deref` calls and ufcs style calls

### CODE REVIEW

**Comment by llogiq** [comment]
I think you need to re-bless some UI test results.

File: clippy_lints/src/dereference.rs [hunk 1]
```diff
@@ -34,76 +36,313 @@ declare_clippy_lint! {
...
+}
+
+struct StateData {
+    /// Span of the top level expression
+    span: Span,
+    /// The required mutability
+    target_mut: Mutability,
+}
+
+enum State {
```
Comment (llogiq) [inline clippy_lints/src/dereference.rs:60]: Why is this a one-variant enum? Do we expect more variants in the future?

File: clippy_lints/src/dereference.rs [hunk 2]
```diff
@@ -34,76 +36,313 @@ declare_clippy_lint! {
...
+}
+
+// Checks whether the parent node is a suitable context for switching from a deref method to the
+// deref operator.
+fn is_linted_explicit_deref_position(parent: Option<Node<'_>>, child_id: HirId) -> bool {
+    let parent = match parent {
+        Some(Node::Expr(e)) => e,
+        _ => return true,
+    };
+    match parent.kind {
```
Comment (llogiq) [inline clippy_lints/src/dereference.rs:250]: Don't we have a utils function for this?

File: clippy_lints/src/redundant_deref.rs [hunk 3]
```diff
@@ -0,0 +1,63 @@
+// use crate::utils::{get_parent_expr, snippet_with_applicability, span_lint_and_sugg};
```
Comment (llogiq) [inline clippy_lints/src/redundant_deref.rs:1]: We don't usually commit commented-out code. Is there a reason for this or is it just a leftover?

File: tests/ui/explicit_deref_methods.stderr [hunk 4]
```diff
@@ -1,67 +1,67 @@
...
+  --> $DIR/explicit_deref_methods.rs:30:19
    |
 LL |     let b: &str = a.deref();
    |                   ^^^^^^^^^ help: try this: `&*a`
    |
    = note: `-D clippy::explicit-deref-methods` implied by `-D warnings`
 
-error: explicit deref_mut method call
-  --> $DIR/dereference.rs:32:23
+error: explicit `deref` method call
```
Comment (llogiq) [inline tests/ui/explicit_deref_methods.stderr:9]: This loses the `_mut` prefix. Is that intentional?

**Review by llogiq — COMMENTED** [review]
In general this seems like mostly a solid improvement! It just needs a small bit of work before we merge it.

File: clippy_lints/src/redundant_deref.rs (hunk 3 above)
Comment (Jarcho) [reply clippy_lints/src/redundant_deref.rs:1]: That is an accident, yes.

File: clippy_lints/src/dereference.rs (hunk 1 above)
Comment (Jarcho) [reply clippy_lints/src/dereference.rs:60]: When this handles both `needless_borrow` and `explicit_auto_deref` there are a few more.

File: clippy_lints/src/dereference.rs (hunk 2 above)
Comment (Jarcho) [reply clippy_lints/src/dereference.rs:250]: Not to go from a node to an expression.

File: tests/ui/explicit_deref_methods.stderr (hunk 4 above)
Comment (Jarcho) [reply tests/ui/explicit_deref_methods.stderr:9]: I didn't keep track of which one was called. I could for the last call in the chain, but that still leaves the mixed case to handle. e.g. `x.deref_mut().deref()`

File: clippy_lints/src/dereference.rs (hunk 2 above)
Comment (llogiq) [reply clippy_lints/src/dereference.rs:243]: Then maybe this should be moved to utils, we might want to use it elsewhere. But that's not a blocker.

**Comment by llogiq** [comment]
Sorry for taking so long. r=me after a rebase.

File: tests/ui/explicit_deref_methods.stderr (hunk 4 above)
Comment (Jarcho) [reply tests/ui/explicit_deref_methods.stderr]: The last call is now tracked.

**Comment by llogiq** [comment]
Thank you! @bors r+

**Pull request merged**